blocks/extensions/plot.py

```diff
@@ -136,14 +132,86 @@ def preexec_fn():
 
-    def __getstate__(self):
-        state = self.__dict__.copy()
-        state['sub'] = None
-        return state
+        self.session = Session(
+            root_url=self.server_url,
+            load_from_config=False)
+        output_server(self.document, session=self.session)
+
+
+class PlotHistogram(SimpleExtension):
+
+    def __init__(self, document, channels, server_url='default', **kwargs):
+        """ Live plot an histogram of mean and std dev of the provided channels
+
+        This extension live plots an histogram of the mean value of each of the
+        provided channels over the monitored dataset. It will also display
+        confidence intervals representing the standard deviation of the channel
+        over the dataset.
+
+        .. warning::
+            This extension assumes that a bokeh server is running at the
+            provided url and won't try to start a server autonomously. To
+            start a bokeh server run `bokeh-server --ip=<ip> --port=<port>`.
+            If you want the server to accept connections from any address use
+            <ip>= 0.0.0.0.
+
+        Parameter
+        ---------
+        document : string
+            The name of the document.
+        channels : list
+            The channels to be displayed as a histogram.
+        server_url : string
+            The url of the server
+        """
+        self.document = document
+        self.channels = channels
+        self.server_url = server_url
 
-    def __setstate__(self, state):
-        self.__dict__.update(state)
-        self._startserver()
-        curdoc().add(*self.p)
+        kwargs.setdefault('after_epoch', True)
+        kwargs.setdefault("before_first_epoch", False)
+        super(PlotHistogram, self).__init__(**kwargs)
+
+    def do(self, which_callback, *args):
+        channels = self.channels
+        log = self.main_loop.log
+        epoch = log.status['epochs_done']
+
+        # create a new document --> a new session
+        self.session = Session(
+            root_url=self.server_url,
+            load_from_config=False)
+
+        output_server(self.document+' (epoch'+str(epoch)+')',
+                      session=self.session, clear=True)
+
+        for key in log.current_row.keys():
+            if key in channels:
+                value = log.current_row[key]
+                kl_mean = value.mean(axis=0)
+                kl_std_dev = value.std(axis=0)
+                nhid = len(kl_mean)
+
+                fig = figure(title=key, plot_width=1000)
+                # display alternating color histograms
+                fig.quad(top=kl_mean[::2], bottom=0, left=range(nhid)[::2],
+                         right=range(nhid+1)[1::2], fill_color='#66E066',
+                         line_color="black")
+                fig.quad(top=kl_mean[1::2], bottom=0, left=range(nhid)[1::2],
+                         right=range(nhid+1)[2::2], fil_color='#66C2FF',
```

Comment:
I'm going to guess that `fil_color` is a typo for `fill_color`? Also you'll rpobably get some pep8 errors on this line